### ISSUE TITLE
k8s-1.7 kubeadm tutorial+romana: kube-dns CrashLoopBackOff

### ISSUE THREAD

**Issue by aalba6675** [romana/romana#168]
I am having a kube-dns CrashLoopBackOff with romana in kubernetes 1.7 with the kubeadm tutorial.
This is a KVM CentOS image with SELinux enforcing set to Permissive. The VM and KVM network are working because I am able to use kubadm/weave the run sock-shop. With romana I get stuck at kube-dns...

```
NAME                               READY     STATUS    RESTARTS   AGE       IP                NODE
etcd-rome0                         1/1       Running   2          33m       192.168.125.110   rome0
kube-apiserver-rome0               1/1       Running   2          33m       192.168.125.110   rome0
kube-controller-manager-rome0      1/1       Running   2          33m       192.168.125.110   rome0
kube-dns-2425271678-dm00z          2/3       CrashLoopBackOff   6          34m       100.112.2.3       rome0
kube-proxy-6f3w2                   1/1       Running   2          34m       192.168.125.110   rome0
kube-scheduler-rome0               1/1       Running   2          34m       192.168.125.110   rome0
romana-agent-zxhl2                 1/1       Running   6          29m       192.168.125.110   rome0
romana-services-3476983861-ddh9n   3/3       Running   20         29m       192.168.125.110   rome0
```

The kubedns container is showing 'Timeout waiting for initialization'.

```
E0706 09:06:38.891385       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Service: Get https://10.96.0.1:443
/api/v1/services?resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout
E0706 09:06:38.897224       1 reflector.go:199] k8s.io/dns/vendor/k8s.io/client-go/tools/cache/reflector.go:94: Failed to list *v1.Endpoints: Get https://10.96.0.1:4
43/api/v1/endpoints?resourceVersion=0: dial tcp 10.96.0.1:443: i/o timeout
```

The romana-* pods do not have any significant errors in their logs.

Both kubeadm and romana are configured to use 10.96.0.0/12 is that correct?
(Weave addon uses 10.32.0.0/12 which is different from the service cluster ip range 😕 )

**Comment by pritesh** [comment]
Hi @aalba6675 can you please post the spec you used for bringing up romana? in the default spec romana uses 100.112.0.0/12 as shown [here](https://github.com/romana/romana/blob/master/containerize/specs/romana-kubeadm.yml#L163). Mainly its different from clusterIP range you mentioned above. Also [here](https://gist.githubusercontent.com/pritesh/5cba3787bb61e34156cde22291c30fe6/raw/6f5aeab66857b13ec51bdcceff4bd42215d867d8/romana-kubeadm.yml) is the latest spec, which may help.

**Comment by aalba6675** [comment]
@pritesh  Your latest spec file (with pani/ images instead of quay.io 1.1.0_ works!
kube-dns pod is up without CrashLoopBackOff - thanks!

Did you fix something between quay.io/romana-*:v1.1.0 and pani/romana-*:2017-06-28?

Here is the diff between the new spec you provided and the one I am using:

```
--- romana.yaml 2017-07-06 08:33:33.882991790 +0000
+++ romana-kubeadm.yml  2017-07-07 00:44:52.007404392 +0000
@@ -119,8 +119,7 @@
         # Uncomment the following line to remove romana persistence data.
         # It should be done when the romana bitmap has changed, or any other
         # config is changed which affects policies/pods/networks, etc.
-        #- "rm -rf /var/lib/romana/*"
-        - "ls -ltr /var/lib/romana/"
+        - "rm -rf /var/lib/romana/*"
         securityContext:
           privileged: true
         volumeMounts:
@@ -157,7 +156,7 @@
         - name: mysql-data
           mountPath: /var/lib/mysql
       - name: romana-services
-        image: quay.io/romana/services:v1.1.0
+        image: pani/romana-services:2017-06-28
         imagePullPolicy: Always
         args:
         - --cidr=100.112.0.0/12
@@ -218,7 +217,7 @@
         effect: NoSchedule
       containers:
       - name: romana-agent
-        image: quay.io/romana/agent:v1.1.0
+        image: pani/romana-agent:2017-06-28
         imagePullPolicy: Always
         args:
         - --cluster-ip-cidr=10.96.0.0/12
@@ -260,4 +259,4 @@
           path: /var/run/romana
       - name: log-path
         hostPath:
-          path: /var/log/romana
+          path: /var/log/romana
\ No newline at end of file
```

**Comment by aalba6675** [comment]
@pritesh Unfortunately, now the second kubelet cannot join and is in CrashLoopBackOff. The log from the romana-agent on the second node:

```
Cluster "kubernetes" set.
User "romana-agent" set.
Context "romana-agent@kubernetes" set.
Switched to context "romana-agent@kubernetes".
Checking topology service is available from romanaok
Internal Error: Host information for '192.168.125.111' had 2 results
```

**Comment by flashvoid** [comment]
Hello, somehow you ended up with 2 hosts sharing one ip in romana database `192.168.125.111` probably just a duplicate of the same host.

You should be able to see this in `romana host list` output.

There are 2 immediate options, you can either connect to etcd and fix your database manually - removing duplicates. 
Or you can delete all the pods, then delete romana `kubectl delete -f romana-kubeadm.yml` and then cleanup wipe `/var/lib/romana` directory before reinstalling romana with `kubectl create -f romana-kubeadm.yml`

There is also a way to do the second option using additional config parameter in `romana-kubeadm.yml` @pritesh can help with with this when he gets online.

**Comment by pritesh** [comment]
Hi @aalba6675 , there are 2 changes in it, First one is that, the cleanup is turned on by default in the new yml file i.e [this one](https://gist.githubusercontent.com/pritesh/5cba3787bb61e34156cde22291c30fe6/raw/6f5aeab66857b13ec51bdcceff4bd42215d867d8/romana-kubeadm.yml)  and it also points to the new pods, where in couple of bug fixes for cni, romana services, etc went in the pods (a numbered release is following soon). Also like @flashvoid mentioned above just reinstalling romana using kubectl delete and then kubectl create commands with this latest yml file should remove any duplicates which are showing up there.

**Comment by aalba6675** [comment]
@flashvoid @pritesh  : when I use `kubeadm join` I am now getting two keys for a new node.
I confirm that the etcd tree does not have that kubelet prior to joining:

```
## existing successful nodes
etcdctl --endpoints http://127.0.0.1:12379 ls /romana/hosts --recursive
/romana/hosts/4
/romana/hosts/1
/romana/hosts/5

## add a new node
ssh rome3 kubeadm join --token 28c00b.491b858471707997 192.168.125.110:6443
# etcdctl --endpoints http://127.0.0.1:12379 ls /romana/hosts --recursive       
/romana/hosts/8
/romana/hosts/1
/romana/hosts/5
/romana/hosts/4
/romana/hosts/7

## two keys created ??
[root@rome0 install]# etcdctl --endpoints http://127.0.0.1:12379 get /romana/hosts/7
{"id":7,"name":"rome3","ip":"192.168.125.113","romana_ip":"100.112.96.0/20","agent_port":9604}
[root@rome0 install]# etcdctl --endpoints http://127.0.0.1:12379 get /romana/hosts/8
{"id":8,"name":"rome3","ip":"192.168.125.113","romana_ip":"100.112.112.0/20","agent_port":9604}
```

Any idea why two keys are created? When I delete one key, then the node can join.

**Comment by flashvoid** [comment]
You should be able to see `listener` process in the process list on master - can you confirm that there is only one?

```
(stas-610)ubuntu@ip-192-168-99-10:~$ ps axf | grep listener
32729 ?        Sl    30:02  |   |       \_ /usr/local/bin/listener -rootURL http://192.168.99.10:9600
```

**Comment by aalba6675** [comment]
It is non-deterministic (whether I get 1 or 2 key(s) for a new node):
- Confirmed that there is a single listener process
- Master node, and first worker joined sucessfully 
- When joining 2nd and 3rd worker I got two keys each

**Comment by chrismarino** [comment]
Hi @aalba6675 if you like, send an email to info@romana and we'll add you to our slack channel. It will get you lower latency responses to these issues.

**Comment by aalba6675** [comment]
@chrismarino  thanks for the suggestion: done!

**Comment by chrismarino** [comment]
@aalba6675 info@romana.io is what I should have indicated....

**Comment by pritesh** [comment]
Fixed in [Release 1.5](https://github.com/romana/romana/releases/tag/v1.5.0)
